ledger-core/network/nds/msgdelivery/controller_parallelism_test.go

```diff
@@ -0,0 +1,273 @@
+					return true
+				default:
+					return false
+				}
+			},
+		}
+	}
+
+	prf3.provider = uniserver.MapTransportProvider(&uniserver.DefaultTransportProvider{},
+		func(provider l1.SessionlessTransportProvider) l1.SessionlessTransportProvider {
+			return l1.MapSessionlessProvider(provider, awaitOutputMapper)
+		},
+		func(provider l1.SessionfulTransportProvider) l1.SessionfulTransportProvider {
+			return l1.MapSessionFullProvider(provider, awaitOutputMapper)
+		})
+
+	numberMessage := 100
+
+	resultsOnServ1 := make(chan string, numberMessage)
+	resultsOnServ2 := make(chan string, numberMessage)
+
+	receiver1 := func(a ReturnAddress, done nwapi.PayloadCompleteness, val interface{}) error {
+		require.True(t, bool(done))
+		s := val.(fmt.Stringer).String()
+
+		resultsOnServ1 <- s
+		return nil
+	}
+
+	receiver2 := func(a ReturnAddress, done nwapi.PayloadCompleteness, val interface{}) error {
+		require.True(t, bool(done))
+		s := val.(fmt.Stringer).String()
+
+		resultsOnServ2 <- s
+		return nil
+	}
+
+	srv1, err := h.createServiceWithProfile(&prf1, receiver1)
+	require.NoError(t, err)
+	srv2, err = h.createServiceWithProfile(&prf2, receiver2)
+	require.NoError(t, err)
+	srv3, err = h.createServiceWithProfile(&prf3, noopReceiver)
+	require.NoError(t, err)
+
+	for i := 0; i < numberMessage; i++ {
+		err := srv3.service.ShipTo(srv2.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("%d", i)}})
+		require.NoError(t, err)
+	}
+
+	<-await
+
+	for i := 0; i < numberMessage; i++ {
+		err := srv3.service.ShipTo(srv1.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("%d", i)}})
+		require.NoError(t, err)
+	}
+
+	select {
+	case res, ok := <-resultsOnServ2:
+		require.True(t, ok)
+		t.Fatalf("unexpected string received:%s", res)
+	default:
+		// channel is empty, this okay
+	}
+
+	checker := func(ch chan string) {
+		res := make(map[string]string)
+
+		for i := 0; i < numberMessage; i++ {
+			val := <-ch
+
+			res[val] = val
+		}
+
+		for i := 0; i < numberMessage; i++ {
+			expVal := fmt.Sprintf("%d", i)
+			_, ok := res[expVal]
+			require.Truef(t, ok, "Not received:%s\n", expVal)
+		}
+	}
+
+	checker(resultsOnServ1)
+
+	out.stop.Store(false)
+	resume <- struct{}{}
+
+	checker(resultsOnServ2)
+}
+
+func TestParallelismReceive(t *testing.T) {
+	h := NewUnitProtoServersHolder(TestLogAdapter{t: t})
+	defer h.stop()
+
+	prf := UnitProtoServerProfile{
+		desFactory: &TestDeserializationStringFactory{},
+	}
+
+	prf1 := prf
+	prf2 := prf
+	prf3 := prf
+
+	pause := make(chan struct{})
+
+	numberMessage := 1000
+
+	receivedFrom1 := make(chan string, numberMessage)
+	receivedFrom3 := make(chan string, numberMessage)
+
+	first := atomickit.NewBool(true)
+
+	receiver2 := func(a ReturnAddress, done nwapi.PayloadCompleteness, val interface{}) error {
+		require.True(t, bool(done))
+		s := val.(fmt.Stringer).String()
+
+		switch {
+		case strings.HasPrefix(s, "1"):
+			if first.Load() {
+				first.Flip()
+				pause <- struct{}{}
```

Comment:
send and receive here looks like an error or misuse

probably send + select with default is needed

Reply:
it won't work, because I read the value from the channel and none message should go through  pause <- struct{}{}

```
err = srv1.service.ShipTo(srv2.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("1%d", 0)}})
	require.NoError(t, err)

<-pause

for i := 1; i < numberMessage; i++ {
	err := srv1.service.ShipTo(srv2.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("1%d", i)}})
	require.NoError(t, err)
}
```

If I will use any other message will go through pause

```
select {
	case pause <- struct{}{}:
	     <-pause
	default:
	}
```

Reply:
ok, my suggestion was wrong
but the given approach is prone for racing anyway. I suggest to revisit it

Reply:
What race can we have? I send one msg and await until the receiver will be called, after that, all messages will go past